### PR TITLE
chore(ci): fail the build when a known large file grows (#918)

### LINKED ISSUE #918: chore(ci): fail the build when a known large file grows (code-line budget)
A CI check that fails when a known large file grows. Not a size limit, and not a claim that any current file is the wrong size: a stop against the *next* line.

## Why

Consolidation has already been done here, in a milestone named for it, and it did not hold. Measured 2026-09-03:

| File | After the refactor that split it | Today | Growth |
|---|---|---:|---:|
| `GameRow.tsx` | **403** ([#340](https://github.com/Stashpeak/SimLauncher/issues/340), v0.9.8 "Refactor & Testing", 2026-05-23) | 844 | **+110 %** |
| `SettingsContext.tsx` | **467** (same issue) | 541 | +16 % |
| `spawn.ts` | 509 when [#530](https://github.com/Stashpeak/SimLauncher/issues/530) was filed, 728 when [#776](https://github.com/Stashpeak/SimLauncher/issues/776) was filed *about that growth* | 1317 | **+159 %** |

`GameRow.tsx` was cut to 403 because it crossed 500, and is now well past the threshold that triggered its own split. [#776](https://github.com/Stashpeak/SimLauncher/issues/776) exists specifically to flag spawn.ts growth and has itself been outgrown by 81 % since it was written.

So the problem is not that consolidation does not happen. **It is that it does not stick, and nothing notices for months.** All three cases above would have been caught on the first added line by a plain upper bound.

## Count code, not lines

Counting total lines would penalise comments, and comments are the best thing about the files in question: `kill.ts` is 43 % comment, `spawn.ts` 38 %, and that documentation is what makes their race conditions maintainable. A total-line budget would charge you for documenting a new one.

Counting code lines also reorders the list usefully:

| File | Code | Total | Comment |
|---|---:|---:|---:|
| `spawn.ts` | 733 | 1317 | 38 % |
| `store.ts` | 700 | 1039 | 17 % |
| `useProfileEditor.tsx` | 637 | 761 | 9 % |
| `GameRow.tsx` | 580 | 844 | 24 % |
| `kill.ts` | 573 | 1145 | 43 % |
| `SettingsContext.tsx` | 482 | 541 | 6 % |

`kill.ts` drops from second to fifth. `useProfileEditor.tsx` rises to third.

## Use the parser, not a regex and not a bare scanner

⚠️ Measured, because the obvious implementation is a trap. `ts.createScanner` with `skipTrivia: false` has no parser context, mis-lexes `/` as the start of a regex literal and swallows whole regions: on `spawn.ts` it reported **105** single-line comments where `grep -cE '^\s*//'` finds **401**. That is worse than the regex it would replace.

`ts.createSourceFile` plus `getLeadingCommentRanges` / `getTrailingCommentRanges` per node is correct, and was verified to reproduce the table above exactly on all 11 files. `typescript` is already a direct devDependency, so this adds nothing.

To be clear about the value: there are **4** JSX comments in all of `src/` today, so the parser buys no accuracy right now. It is there so the count stays right when someone adds `{/* ... */}`, a template string containing `//`, or a regex literal, rather than silently charging them budget for a comment.

## Scope

- Recursively check version-controlled `.ts` and `.tsx` under `src/`.
- A line counts as code if it holds a real token. Blank and comment-only lines do not. A line with code plus a trailing comment is one code line.
- Handle `//`, `/* */`, JSX `{/* */}`, regex literals, template strings containing comment markers, and both CRLF and LF.
- A parse failure fails the check. It must never silently produce a wrong number.
- `scripts/size-budget.json` holds the 300-line threshold, a ceiling per tracked file, and exclusions each carrying a human reason.
- The 11 files currently at or above 300 code lines get today's count as their ceiling.
- `src/renderer/src/components/icons/index.tsx` is excluded: 374 code lines at 1 % comment, a declarative SVG barrel. It is data, not complexity.
- Paths normalised to `/` so Windows and Linux agree.
- `npm run size:check` never writes. Non-zero exit on any violation, and all violations printed at once with path, current, budget, delta and the fix.
- Fails when a tracked file exceeds its ceiling, when a new file crosses 300 without an entry, or when an entry points at a path that no longer exists.
- A tracked file stays tracked after dropping below 300, so a refactored file cannot quietly climb back to the threshold.
- Raising **and** lowering a ceiling are the same act: edit one number, in the same PR, visible in review.
- Runs in the existing `build` job in `ci.yml`. No new workflow.
- One paragraph in `AGENTS.md`. No other documentation.

## Explicit non-goals

- Not a complexity metric and not an architecture judgement.
- Does not apply to tests, build output or dependencies.
- Does not forbid large files or propose splits.
- Does not replace code review or the planned refactors.
- **Today's ceilings are not an endorsement.** 844 code lines in `GameRow.tsx` is not fine; the ceiling only stops 845. Lowering them is the job of the refactor issues ([#530](https://github.com/Stashpeak/SimLauncher/issues/530), [#775](https://github.com/Stashpeak/SimLauncher/issues/775), [#776](https://github.com/Stashpeak/SimLauncher/issues/776), [#396](https://github.com/Stashpeak/SimLauncher/issues/396)), whose acceptance should include dropping the relevant number.

## Rejected, with the trigger to revisit

**A downward ratchet plus `npm run size:update`** (fail when a tracked file falls more than 5 % under its ceiling, so refactor gains are banked automatically). Designed in full, then dropped as unproven need.

Against the three failures above it adds nothing: every one is caught by the upper bound on the first added line. It defends only against a case never observed here, where someone refactors, forgets to lower the number, and the file regrows within the old ceiling. Its costs are real and immediate: false CI failures on any PR that deletes a few lines from a tracked file, plus a second command whose entire specification is about not letting it launder a red build green.

**Revisit when** a refactored file demonstrably re-inflates because its ceiling was left high. Then it is an hour's work against evidence.

## Acceptance

- [ ] `npm run size:check` runs in the existing `build` job and exits non-zero on any violation
- [ ] Counting uses the `typescript` parser; no regex or standalone-scanner counting
- [ ] `scripts/size-budget.json` carries the threshold, the 11 current ceilings, and `icons/index.tsx` excluded with its reason
- [ ] Over-ceiling, new-file-over-threshold, and stale-path each fail with a readable message; multiple violations report together
- [ ] Tests cover: parser handling of line, block and JSX comments plus regex and template-string edge cases; equal-to-budget passes and one over fails; a new file over 300 without an entry fails; exclusions and Windows path normalisation; multiple violations reported at once
- [ ] One paragraph in `AGENTS.md`
- [ ] Full gate suite green, and the check runs offline in well under a second

### PR DESCRIPTION
**Pull request by Shieldxx**
## Summary

`npm run size:check` runs in the existing `build` job and fails when a file that is already large gets larger. Ceilings live in `scripts/size-budget.json`.

**This is not a size limit and the committed numbers are not an endorsement.** 569 code lines in `GameRow.tsx` is not fine; the ceiling only stops 570. Lowering them belongs to [#530](https://github.com/Stashpeak/SimLauncher/issues/530), [#775](https://github.com/Stashpeak/SimLauncher/issues/775), [#776](https://github.com/Stashpeak/SimLauncher/issues/776) and [#396](https://github.com/Stashpeak/SimLauncher/issues/396).

### Why

Consolidation has been done here before, in a milestone named for it, and it did not hold:

| File | After the refactor that split it | Today |
|---|---|---:|
| `GameRow.tsx` | **403** ([#340](https://github.com/Stashpeak/SimLauncher/issues/340), v0.9.8 "Refactor & Testing") | 844 |
| `spawn.ts` | 509 at [#530](https://github.com/Stashpeak/SimLauncher/issues/530), 728 at [#776](https://github.com/Stashpeak/SimLauncher/issues/776), which was filed *about that growth* | 1317 |

Nobody noticed for months. All of it is caught by a plain upper bound on the first added line.

### Counts code, not lines

`kill.ts` is 43 % comment and `spawn.ts` 38 %, and that documentation is why their race conditions stay maintainable. A total-line budget would charge for writing it. **A comment is always free to add; logic is not.**

### The parser, and the trap in it

⚠️ Worth recording because the obvious implementation is wrong. `ts.createScanner` has no parser context, reads `/` as the start of a regex literal and swallows whole regions: on `spawn.ts` it reported **105** single-line comments where the file has **401**. That is worse than the regex it replaces. `ts.createSourceFile` with `getLeadingCommentRanges` is correct.

Parsing then surfaced a case neither approach handles for free: `{/* why */}` is a `JsxExpression` whose braces are **real tokens**, so it counted as code and adding a JSX comment would have cost budget. Claimed explicitly. This moved `GameRow.tsx` from 580 to 569 and `App.tsx` from 404 to 399.

### Rejected

A downward ratchet plus `size:update`, designed in full and dropped as unproven need: it catches none of the three observed failures, and it costs false CI failures on any PR that deletes a few lines from a tracked file. The trigger to revisit is recorded on [#918](https://github.com/Stashpeak/SimLauncher/issues/918).

## Smoke check

- [x] **Needs no manual check.**  Reason: no file under `src/` is touched, so the app and the shipped bundle are byte-for-byte unaffected. The change is repo tooling plus one npm script, one CI step and one `AGENTS.md` section.
- [ ] **Needs a manual check.**  What to do, and what a correct result looks like:

## Test plan

- [x] `npm run format:check`
- [x] `npm run lint`
- [x] `npm run typecheck`
- [x] `npm run size:check`
- [x] `npm run test` (935 passed, 21 new)
- [x] `npm run build`
- [x] **Proven red, not just green.** Adding one line of code to `GameRow.tsx` exits 1 with `570 code lines, budget 569 (+1)`. Adding two comment lines to the same file stays green and exits 0.
- [x] Also exercised by hand: a new tracked file over 300 lines with no entry, a stale budget entry, a stale exclusion, and a path both budgeted and excluded. All fail, and all violations report together.

## Checklist

- [x] `npm run build` passes
- [x] `npm run typecheck` passes
- [x] `npm run format:check` passes
- [x] Tested manually — the red/green pair above

Closes #918

Closes #918

## Summary by CodeRabbit

* **New Features**
  * Added automated code-size budget checks to the build process.
  * Large source files are now monitored against configured code-line limits, with clear build feedback when limits are exceeded.

* **Documentation**
  * Added contributor guidance explaining code-size thresholds, exclusions, and budget updates.

* **Tests**
  * Added coverage for line counting, budget validation, configuration errors, and reported violations.

### CODE REVIEW

File: scripts/countCodeLines.mjs [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    for (const range of found) {
+      const key = `${range.pos}:${range.end}`
+      if (seen.has(key)) continue
+      seen.add(key)
+      ranges.push(range)
+    }
+  }
+
+  const visit = (node) => {
+    add(ts.getLeadingCommentRanges(text, node.getFullStart()))
```
[the remark this answers is not in the document]
Comment (Shieldxx) [reply scripts/countCodeLines.mjs]: Confirmed and fixed in `aed8bed`. I reproduced it before changing anything: a comment on its own line before a closing token was counted as code in a block body, an object literal, an array literal and an `if` body, all four.

Fixed more broadly than reported. Rather than adding the three container closing tokens, `collectCommentRanges` now walks `node.getChildren(sourceFile)` instead of `forEachChild`, which includes punctuation tokens, so every container is covered rather than only the ones that happened to get named. It also subsumes the `endOfFileToken` special case, which is now gone.

Regression cases added for all four shapes plus a comment on the last line of a file. No committed budget number moved, so no source file currently carries the pattern, which is why it went unnoticed.

Worth saying: you and CodeRabbit found this independently, and it is the one finding here that went against the stated design goal rather than merely being untidy.

File: AGENTS.md [hunk 2]
```diff
@@ -46,6 +46,14 @@ Electron desktop app for simracing enthusiasts — React + TypeScript + Tailwind
 - Doc-comment public/exported items (TS JSDoc) where applicable. English only.
 - Applies to AI agents too.
 
+## Code-line budget
+
+`npm run size:check` runs in the required `build` job and fails when a file that is already large gets larger. Ceilings live in [`scripts/size-budget.json`](scripts/size-budget.json); any `src/` file crossing 300 lines of code needs an entry there, or an exclusion with a written reason.
+
+It counts **code** only, so blank and comment-only lines are free. Add the comment. `kill.ts` is 43% comment and that documentation is why its race conditions are maintainable; a budget that charged for it would push exactly the wrong way.
+
+**Raising a ceiling is allowed — edit the number in the same PR.** The gate does not forbid growth, it makes growth appear in a diff. The current numbers are not an endorsement: `GameRow.tsx` was cut to 403 lines by #340 in the 0.9.8 refactor milestone and is over 800 today, and `spawn.ts` went 509 → 728 → 1317 while two issues sat open about that growth. Nobody noticed for months. A refactor PR should lower its file's number; nothing forces it to, and that omission was a deliberate call recorded on #918.
```
[the remark this answers is not in the document]
Comment (Shieldxx) [reply AGENTS.md]: Fair, and that was scope creep against my own acceptance criterion, which said one paragraph. Condensed to one in `aed8bed`, keeping the ceiling location, the code-not-lines rule, the "raising is allowed, do it in the diff" instruction and the GameRow example, and leaving the rest on the issue.

File: scripts/checkSizeBudget.mjs [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+      violations.push(
+        `stale exclusion: ${file}\n` +
+          `  no such tracked file under src/\n` +
+          `  fix: remove the entry, or correct the path if the file moved`
+      )
+    }
+  }
+
+  for (const file of [...present].sort()) {
+    if (Object.hasOwn(exclusions, file)) continue
```
[the remark this answers is not in the document]
Comment (Shieldxx) [reply scripts/checkSizeBudget.mjs]: Right, and fixed in `aed8bed`. `Object.hasOwn` was checking for the key and ignoring the value, so `"src/huge.ts": ""` would have bought a silent exemption from the written reason this gate exists to require.

An exclusion is now honoured only when its reason is a non-empty string after trimming. Anything else reports `is excluded with no reason` and the file is measured normally. Parameterised regression cases cover empty string, whitespace, `null` and a non-string.

This was the sharpest of the five: it is the one way the config could have been weakened while the diff still looked like nothing had happened.

File: scripts/checkSizeBudget.mjs [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+      )
+      continue
+    }
+
+    const code = codeLinesFor(file)
+
+    if (budget === undefined) {
+      // A file only needs an entry once it crosses the threshold. Below it,
+      // files are unlisted and unconstrained.
+      if (code >= threshold) {
```
[the remark this answers is not in the document]
Comment (Shieldxx) [reply scripts/checkSizeBudget.mjs]: Holding the behaviour, fixing the wording instead.

`thresholdCodeLines: 300` names the size at which the rule applies, so a file sitting exactly at 300 is exactly as large as the size we said we care about, and tracking it is the stricter reading. Where the wording is ambiguous I would rather a gate be strict than loose.

You are right that the prose disagreed with the code, and that is the real defect here. `AGENTS.md` now says "at or above 300 lines of code" in `aed8bed`, and the test pinning the inclusive boundary stays.

File: scripts/countCodeLines.mjs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Shieldxx) [reply scripts/countCodeLines.mjs]: Confirmed and fixed in `aed8bed`. Our own review bot raised the same gap independently, so it got the fuller treatment.

Rather than adding the three named containers, `collectCommentRanges` now walks `node.getChildren(sourceFile)` instead of `forEachChild`, so every closing token is covered rather than only the reported three, and the separate `endOfFileToken` call is no longer needed. Regression cases added for block, object and array, plus an `if` body and a comment on the final line of a file.

**Comment by Shieldxx** [comment]
@codex review

**Comment by Shieldxx** [comment]
@coderabbitai full review

**Comment by Shieldxx** [comment]
@codex review

File: scripts/checkSizeBudget.mjs [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+    cwd: repoRoot,
+    encoding: 'utf8',
+    maxBuffer: 32 * 1024 * 1024
+  })
+
+  return output
+    .split('\0')
+    .filter(Boolean)
+    .map(toPosix)
+    .filter((file) => file.endsWith('.ts') || file.endsWith('.tsx'))
```
[the remark this answers is not in the document]
Comment (Shieldxx) [reply scripts/checkSizeBudget.mjs]: Confirmed, and the measurement holds: `App.css` is 941 physical lines at `7984c66`, 130 blank and 97 comment, so roughly 714 code lines. Well past the 300 the same sentence sets.

I checked whether CSS was the only gap. It is. The only other non-TypeScript file under `src/` is one `.html` entry point below 100 lines, so closing this means covering CSS and covering CSS means covering exactly one file today.

Splitting it rather than fixing it here, and the split is deliberate. In `86487b0` the AGENTS.md sentence now says *TypeScript file* and names the gap out loud, so the documented rule matches what the gate does instead of overclaiming. #921 tracks actually measuring CSS, with the growth data behind the priority call: `App.css` went 726 to 941 across the last five releases, about 30 percent, against `spawn.ts` at 509 to 1317, about 159 percent, over the same span. Real growth, the slowest of the large files.

The reason it is not in this PR is risk, not effort. CSS lexing is genuinely easier than TypeScript, one comment form and no regex-literal ambiguity, but this counter has already produced two comment-detection bugs and the fix for the second one has not been reviewed yet. Adding a second lexer underneath an unreviewed fix is how the third one arrives.

**Comment by Shieldxx** [comment]
@codex review

File: scripts/countCodeLines.mjs [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+
+  visit(sourceFile)
+
+  return ranges
+}
+
+/** Whether the line still holds non-whitespace once every comment on it is blanked out. */
+function hasCodeOutsideComments(lineText, lineStart, ranges) {
+  const lineEnd = lineStart + lineText.length
+  const chars = [...lineText]
```
[the remark this answers is not in the document]
Comment (Shieldxx) [reply scripts/countCodeLines.mjs]: Correct, and reproduced exactly as written. Fixed in `e2cf3af` by building the mask with `split('')`, which splits on UTF-16 code units and so matches what the parser hands back.

Worth recording that my first attempt to reproduce this failed, and the failure was mine rather than a problem with the report. I probed `/*<emoji>*/const x = 1`, which passes: the drift blanks one extra character, `const` loses its `c`, and the line still holds plenty of non-whitespace, so the count never flips. Your example uses a one-character tail for a reason. Measured on the pre-fix code:

| Source | code lines | correct |
|---|---|---|
| `/*<emoji>*/x` | 0 | 1 |
| `/*<emoji><emoji>*/xy` | 0 | 1 |
| `/*<emoji><emoji>*/xyz` | 1 | 1 |
| `/*<accented e>*/x` | 1 | 1 |

So the rule is that N astral characters eat N characters past the comment, and the count only flips once that reaches the end of the line. Anything with a longer tail hides it, which is what made my weaker probe agree with the code.

Four regression cases plus the non-astral control are in `tests/ci/sizeBudget.test.ts`. Verified red before green: with the mask reverted to `[...lineText]`, exactly the two short-tail cases fail with `expected +0 to be 1` and the other two pass either way, so the tests are pinning the bug rather than the behaviour.

One thing that did not change the decision. No file under `src/` currently contains an astral character, all 120 scanned, so nothing is being miscounted today. Fixed regardless: this file's own contract is that a wrong number which looks fine is the one outcome a budget gate must never produce, and a line that can silently read as comment-only is exactly that.

**Comment by Shieldxx** [comment]
@coderabbitai full review

**Comment by Shieldxx** [comment]
Both nitpicks addressed in `36daee7`, one fixed and one half held. They sat in the review body rather than as threads, hence a top-level reply.

**`parseDiagnostics` guard: fixed, and it was the sharper of the two.** You are right that `?? []` made an internal-API disappearance indistinguishable from a clean parse. A TypeScript upgrade that renamed the property would have turned the throw underneath it into a silent pass, and an unparseable file would then have been counted rather than rejected, which is the exact outcome the function's own docstring rules out. It now rejects anything that is not an array, so the gate fails closed on an API change.

Verified by deleting the property at runtime rather than by reading the diff: the control run counts normally, and with the property gone the call throws `could not be parsed (TypeScript no longer exposes parseDiagnostics)`. There is no unit test, and deliberately so: reaching that branch means stubbing the `typescript` module itself, which buys less than it costs for a branch whose untested direction is to throw.

I did not move to a public diagnostics API. `getPreEmitDiagnostics` needs a full `Program`, which would turn a per-file counter into a whole-project type-check for no gain in what it measures. Keeping the internal property and failing loudly when it vanishes is the cheaper half of your suggestion, and it closes the same hole.

**Test comments: one removed, one kept.** The comment above `toHaveLength(3)` is gone. You are right that it listed three violations that the three assertions directly beneath it already name, so it restated the code.

Holding the one at line 65. `// Six lines, one of them a JSX comment.` is what justifies the magic `5`, and nothing else in that test says the fixture has six lines: the source is a single-line string literal, so a reader would otherwise have to count `\n` escapes by eye to check the assertion is right. That is arithmetic a reader cannot see, which reads to me as a WHY rather than a WHAT under the same guideline you cited. Happy to be overruled if you read the rule more strictly.

**Comment by Shieldxx** [comment]
@codex review

File: scripts/countCodeLines.mjs [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+    const first = syntactic[0]
+    const message = ts.flattenDiagnosticMessageText(first.messageText, ' ')
+    throw new Error(`${filePath}: could not be parsed (${message})`)
+  }
+
+  const ranges = collectCommentRanges(text, sourceFile)
+
+  // Split on the three line endings rather than trusting one, then drop the
+  // empty tail a trailing newline produces so counts match `wc -l`.
+  const lines = text.split(/\r\n|\r|\n/)
```
[the remark this answers is not in the document]
Comment (Shieldxx) [reply scripts/countCodeLines.mjs]: Correct, and reproduced. Fixed in `b4db32c`.

Measured before the fix, comparing our split against the parser's own line map:

| Source | ts line starts | our lines | code counted | correct |
|---|---|---|---|---|
| three statements, LF | 4 | 3 | 3 | 3 |
| three statements, U+2028 | 4 | 1 | **1** | 3 |
| three statements, U+2029 | 4 | 1 | **1** | 3 |
| two statements and a comment, U+2028 | 4 | 1 | **1** | 2 |

So it is worse than a miscount. The offsets used for masking come from `getPositionOfLineAndCharacter`, which is the parser's line map, while the lines being masked came from a different split. The two disagreed, and every line after the first separator was indexed against the wrong start.

The split now includes both separators. Three regression cases, verified red before green: with the split reverted to the three common terminators, exactly those three fail and the other 37 pass.

One deliberate detail. The separators are written as the escapes `\u2028` and `\u2029` in both the counter and the tests, never as the literal characters. My first attempt pasted the real code points into the regex, which works and is exactly the wrong thing to leave behind: an invisible line terminator sitting in source, in the fix for a bug about invisible line terminators. It would also not survive the next editor or formatter that decides to normalise it.

No file under `src/` uses either separator today, all 120 scanned, so nothing was being miscounted. Fixed for the same reason as the UTF-16 drift you found earlier: the contract in this file's docstring is that it must never return a wrong number that looks fine, and a file reporting five lines instead of five hundred looks entirely fine.

That is three real defects in this counter across two reviews, all three in the same seam, which is where our idea of a line or a character stops matching the parser's.

**Comment by Shieldxx** [comment]
@codex review
